### PR TITLE
Add "argon" variant of song progress display

### PR DESCRIPTION
**Pull request by ItsShamed**
Related:
- #18888 

Depends on:
- [x] #22143 (rework needs to be done once this gets merged, reviews about `SegmentedGraph` go there)
- [x] https://github.com/ppy/osu/pull/22273

Also addresses #21814 in some way, as it will no longer distract the player during gameplay.

I tried to implement the argon variant of the song progress

image

https://user-images.githubusercontent.com/39100084/210313679-e2000646-6f28-4c6e-95c4-81a0196a0328.mp4

I made some assumptions regarding UX since I don't know what exactly are the directions about this, any suggestion is welcomed.

### CODE REVIEW

**Comment by peppy** [comment]
Looks like a pretty amazing start! I do think that the density should be visible even from time in the past, rather than having it fully opaque white. Maybe have a play with that and see if you can dim it slightly and still have the density view above the progress.

**Comment by peppy** [comment]
cc @arflyte please view the video in the opening post and provide any feedback (also see my comment above).

File: osu.Game/Screens/Play/HUD/ArgonSongProgress.cs
Comment (ItsShamed) [inline osu.Game/Screens/Play/HUD/ArgonSongProgress.cs:59]:
```suggestion
                graph = new ArgonSongProgressGraph
                {
                    Name = "Difficulty graph",
                    Origin = Anchor.BottomLeft,
                    Anchor = Anchor.BottomLeft,
                    RelativeSizeAxes = Axes.X,
                },
```
Cause of test failures, should be probably enough. I'm not home currently so I cannot make changes right now.

**Comment by ItsShamed** [comment]
I think this whole test class makes no sense, since pretty much everything is tested in the tests for `SegmentedGraph`. @peppy thoughts on this?
I will remove it locally so that I can experiment more on my side, let me know if I should do anything.

**Comment by ItsShamed** [comment]
> Maybe have a play with that and see if you can dim it slightly and still have the density view above the progress.

With additive blending applied on the graph I get this.

image

**Comment by peppy** [comment]
I think that looks fine, let's go with that.

You'll need to look into test and code quality failures on CI as well.

**Comment by arflyte** [comment]
I love what I'm looking at so far :D

**Comment by peppy** [comment]
I've fixed one issue, but will leave the rest to you.

Also, it looks like masking isn't being applied correctly:

osu! 2023-01-17 at 09 07 41

**Comment by ItsShamed** [comment]
Oh you were faster than me 😅
I was going to fix the tests anyways, and masking is a known issue (fixed in next pending commits)

**Comment by peppy** [comment]
I've made an initial cleanup pass on this PR. Please check each of my commits to see what was required to get things up-to-scratch.

But I do wonder: how were you testing this? Because the `SongProgress` test scene doesn't work properly with enabling / disabling seek or anything, and it also cuts off the edges of the display and doesn't show any density data. I'll fix that up, but please make sure that you use test scenes for developing components, not the actual game.

**Comment by ItsShamed** [comment]
I did used the `SongProgress` test scene, although there was no "beatmap density" data fed into any of the `SongProgress`es components which is why I mostly used the game as a test for that.
But yeah, I should use tests more often and should have taken the time to do one at least before opening this PR. On top of that I have no reason to not make one since I did make one for #22143 😅
My bad

File: osu.Game/Screens/Play/HUD/ArgonSongProgress.cs
Comment (ItsShamed) [inline osu.Game/Screens/Play/HUD/ArgonSongProgress.cs:33]:
```suggestion
        // as this component is not contained within the `FrameStabilityContainer` and won't see the dependency.
```

**Review by ItsShamed — COMMENTED** [review]
`ClicksPerSecondCalculator` copy-pasta?

File: osu.Game/Screens/Play/HUD/ArgonSongProgress.cs
Comment (peppy) [reply osu.Game/Screens/Play/HUD/ArgonSongProgress.cs]: It's fine that code is now gone with further refacoring.

**Comment by peppy** [comment]
I've finished my pass on this PR. Will request one more team review as I've changed a lot.

There's probably still work that can be done on this, but I'd like to get this in sooner rather than nitpicking, so I'd recommend anyone re-reviewing this applies changes themselves and goes ahead with the merge, bar anything major being wrong.

**Review by peppy — DISMISSED** [review]

**Review by peppy — DISMISSED** [review]

**Review by frenzibyte — COMMENTED** [review]
I got interrupted as I was reviewing this, but it looked mostly fine to me (minus one issue I noticed when hiding HUD where the song progress bar doesn't hide at all, which is unrelated to this PR and more focused on `SegmentedGraph` drawnodes instead). Need a second review to ensure everything is good.

**Review by bdach — APPROVED** [review]

**Pull request merged**
